index.html

```diff
@@ -430,6 +440,15 @@ <h2><dfn>disconnect()</dfn> method</h2>
         <a>relevant global object</a>, and also empty <a>context object</a>'s
         <a>observer buffer</a>.</p>
       </section>
+      <section>
+        <h2><dfn>supportedEntryTypes</dfn> attribute</h2>
```

Comment:
Hmm I believe that's unnecessary right now: I doubt any web developer or browser would complain about the meaning of this being unclear. The benefit of registering to me would be unified behavior in terms of the buffers of each entry type. But that is currently not unified (i.e. our buffers have some differences depending on the entry type). So I'd see the value of the registration idea, but only if/when the buffers of all entry types behave the same way, so we could describe the entry types' buffer behavior in PerformanceTimeline. Does this make sense?

Reply:
IMO, adding this without such hooks would add more debt and under-defined concepts to the spec. As I commented below, "unsupported types" is already an under-defined concept, but I don't know if we want to add to its usage.

Reply:
I disagree because in my view there is nothing to be defined about what "user agent support" means. It's also unclear when such hooks would be executed (it needs to be very early on to ensure this method never returns an incorrect value). So IMO the hooks could be added when there is some additional incentive to add them, not to define what supported entry types means. I do agree that we can improve the wording so that it is clear what we mean by these supported entry types, but I believe the hooks are not needed for this. To further argument my case with an example, I'd like to point out to this spec which exposes a similar attribute:
https://w3c.github.io/push-api/#dom-pushmanager-supportedcontentencodings

Reply:
Oh ok I think I see some benefit of having a 'hook' now: if you're implementing say PaintTiming, you'd like to know that you need to update supportedEntryTypes accordingly. Thus, the PaintTiming spec should somehow mention the need for the user agent to update the array. Does this sound like the benefit that was being considered?

If that's the case, we can either a) explicitly call it out in each spec ('the user agent must not include "paint" in PerformanceObserver.supportedEntryTypes") or b) define a hook to "register entry type" which would do that. Ilya's initial sketch included buffer behavior, but like I said they currently behave differently so at this point we probably don't want to include buffers in the hook.

If any of those are resonable, what would be better, a) or b)? a) is simpler in the short term, b) means we can hook more shared implementation details later if needed.

Reply:
> Oh ok I think I see some benefit of having a 'hook' now: if you're implementing say PaintTiming, you'd like to know that you need to update supportedEntryTypes accordingly. Thus, the PaintTiming spec should somehow mention the need for the user agent to update the array. Does this sound like the benefit that was being considered?

Exactly. We want to make sure the list remains up-to-date as new entry types are added, and the cleanest way to do that is to burden the dependent specs to update it.



> If that's the case, we can either a) explicitly call it out in each spec ('the user agent must not include "paint" in PerformanceObserver.supportedEntryTypes") or b) define a hook to "register entry type" which would do that. Ilya's initial sketch included buffer behavior, but like I said they currently behave differently so at this point we probably don't want to include buffers in the hook.

b) sounds like the better alternative, as it doesn't require the different dependent specs to be aware of each other. I agree we can start with simple entry type registration (the buffering related hooks are a different beast and we should tackle them separately).

Reply:
Yeah, I don't see how we can provide a hook for other specifications since we need to specify the order in which types are enumerated. We can't let multiple specifications specify the type without having a total order across them.

Reply:
Why do we need to specify order? If the order needs to be predictable, can't we sort the list in a predictable way?

Reply:
Oh, that's a good point. We should definitely sort the type names if we were to provide hooks.

Reply:
As a thought experiment.. Instead of defining a "text hook" where upstream specs register types, could we instead just define this in IDL via [partial dictionary](https://heycam.github.io/webidl/#dfn-partial-dictionary) that each spec extends? Concretely..

1. PerfTimeline defines `entryTypes` which is an empty dictionary
1. Upstream specs provide partial's to extend said dictionary

Crazytalk?

Reply:
Why do we wan to use IDL for this? I don't think that makes much sense.

Reply:
Registrations in prose are hard to track across specs and error prone due to nature of prose. With IDL, the registration is implicit — we don't need to define any hooks and such — and we can scan and pull in all the IDL declarations across specs, etc. Not saying we must do it this way, but worth considering.

Reply:
I think the array I'm proposing is a bit cleaner. Dictionaries contain both keys and values, whereas here we would just need the keys (the strings). Also, WPT tests could help with making sure that the implementor updates `supportedEntryTypes` when needed.